types/src/sample.rs

```diff
@@ -2,45 +2,153 @@ use std::mem::size_of;
     pub fn new(
         axis_type: AxisType,
         index: usize,
-        dah: &DataAvailabilityHeader,
+        eds: &ExtendedDataSquare,
         block_height: u64,
     ) -> Result<Self> {
-        let square_len = dah.square_len();
+        let square_len = eds.square_len();
 
         let (axis_index, sample_index) = match axis_type {
             AxisType::Row => (index / square_len, index % square_len),
             AxisType::Col => (index % square_len, index / square_len),
         };
 
+        let shares = eds.axis(axis_type, axis_index)?;
+
+        let mut tree = Nmt::with_hasher(NamespacedSha2Hasher::with_ignore_max_ns(true));
+
+        // XXX: are erasure coded shares correctly prefixed with parity namespace?
+        for s in &shares {
+            tree.push_leaf(s.data(), *s.namespace())
+                .map_err(Error::Nmt)?;
+        }
+
+        let proof = NmtNamespaceProof::PresenceProof {
+            proof: tree.build_range_proof(sample_index..sample_index + 1),
+            ignore_max_ns: true,
+        };
+
+        let sample_id = SampleId::new(index, square_len, block_height)?;
+
+        Ok(Sample {
+            sample_id,
+            sample_proof_type: axis_type,
+            share: shares[sample_index].clone(),
+            proof: proof.into(),
+        })
+    }
+
+    /// Validate sample with root hash from ExtendedHeader
+    pub fn validate(&self, hash: Hash) -> Result<()> {
+        let namespaced_hash = NamespacedHash::from_raw(hash.as_ref())?;
+        self.proof
+            .verify_range(&namespaced_hash, &[&self.share], *self.share.namespace())
+            .map_err(Error::RangeProofError)
+    }
+}
+
+impl Protobuf<RawSample> for Sample {}
+
+impl TryFrom<RawSample> for Sample {
+    type Error = Error;
+
+    fn try_from(sample: RawSample) -> Result<Sample, Self::Error> {
+        let Some(proof) = sample.sample_proof else {
+            return Err(Error::MissingProof);
+        };
+
+        let sample_id = SampleId::decode(&sample.sample_id)?;
+        let share = Share::from_raw(&sample.sample_share)?;
+        let sample_proof_type = u8::try_from(sample.sample_type)
+            .map_err(|_| Error::InvalidAxis(sample.sample_type))?
+            .try_into()?;
+
+        Ok(Sample {
+            sample_id,
+            sample_proof_type,
+            share,
+            proof: proof.try_into()?,
+        })
+    }
+}
+
+impl From<Sample> for RawSample {
+    fn from(sample: Sample) -> RawSample {
+        let mut sample_id_bytes = BytesMut::new();
+        sample.sample_id.encode(&mut sample_id_bytes);
+        let sample_proof = RawProof {
+            start: sample.proof.start_idx() as i64,
+            end: sample.proof.end_idx() as i64,
+            nodes: sample.proof.siblings().iter().map(|h| h.to_vec()).collect(),
+            leaf_hash: vec![], // this is an inclusion proof
+            is_max_namespace_ignored: true,
+        };
```

Comment:
don't we have this implemented somewhere already?

Reply:
Indeed we have it in the `namespace_proof`, replacing with `into()` - thanks!